walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,92 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> ret = new ArrayList<>();
+    if (request.getDuration() <= 0) {
+      return ret;
+    }
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return ret;
+    }
+    List<TimeRange> occupiedBlocks = recordOccupiedBlocks(events, request.getAttendees());
+    ret = complement(TimeRange.WHOLE_DAY, occupiedBlocks);
+    ret.removeIf(tr -> tr.duration() < request.getDuration());
+    return ret;
+  }
+
+  /* Return list of blocks occupied by attendees, unordered */
+  private List<TimeRange> recordOccupiedBlocks(Collection<Event> events, Collection<String> names) {
+    List<TimeRange> occupiedBlocks = new ArrayList<>();
+    for (Event event : events) {
+      if (isOccupied(event, names)) {
+        occupiedBlocks.add(event.getWhen());
+      }
+    }
+    return occupiedBlocks;
+  }
+
+  /* Return true if anyone in the targetNames is involved in the event */
+  private boolean isOccupied(Event event, Collection<String> targetNames) {
+    for (String targetName : targetNames) {
+      if (event.getAttendees().contains(targetName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /* Return a list of the available TimeRanges */
+  private List<TimeRange> complement(TimeRange wholeDay, List<TimeRange> mergedTimeRanges) {
+    Stack<TimeRange> complStack = new Stack<>();
+    Stack<TimeRange> rawStack = new Stack<>();
+
+    // TimeRanges are added in the order that they are returned by mergedTimeRanges's iterator
+    // The earliest (by start time) TimeRange is on top.
+    mergedTimeRanges.sort(TimeRange.ORDER_BY_START.reversed());
```

Comment:
Yes I totally agree that modifying the input of a method isn't a good practice. I don't know if sorting the time ranges in the main function is a good idea though, because it isn't obvious to me why sorting is necessary in `query()` without looking into the `complement()`.

Reply:
I still think its not that good because by reading the block of main code without looking at the comment for the helper method, I wouldn't know if its modifying. If I see sorting in the main method I would think "I guess the helper requires it to be sorted" which is better than not knowing its getting modified. You should also change the method parameter to indicate the input should be sorted.